components/inbound-endpoints/org.wso2.carbon.inbound.endpoint/src/main/java/org/wso2/carbon/inbound/endpoint/protocol/nats/StreamingListener.java

```diff
@@ -0,0 +1,215 @@
+import org.apache.synapse.SynapseException;
+import org.wso2.carbon.inbound.endpoint.protocol.nats.management.NatsEndpointManager;
+
+import java.io.IOException;
+import java.time.Duration;
+import java.util.Properties;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * Streaming listener class which uses NATS streaming connection to receive messages.
+ */
+public class StreamingListener implements NatsMessageListener {
+
+    private static final Log log = LogFactory.getLog(StreamingListener.class.getName());
+    private String subject;
+    private NatsInjectHandler injectHandler;
+    private Properties natsProperties;
+    private StreamingConnection connection;
+    private Subscription subscription;
+
+    public StreamingListener(String subject, NatsInjectHandler injectHandler, Properties natsProperties) {
+        this.subject = subject;
+        this.injectHandler = injectHandler;
+        this.natsProperties = natsProperties;
+    }
+
+    /**
+     * Create the connection to the NATS Streaming server.
+     *
+     * @return boolean value whether connection is created.
+     */
+    @Override public boolean createConnection() throws IOException, InterruptedException {
+        if (connection == null) {
+            String natsStreamingUrl = natsProperties.getProperty(NatsConstants.NATS_STREAMING_URL);
+            String natsStreamingClientId = natsProperties.getProperty(NatsConstants.NATS_STREAMING_CLIENT_ID);
+            String natsStreamingClusterId = natsProperties.getProperty(NatsConstants.NATS_STREAMING_CLUSTER_ID);
+            String connectWait = natsProperties.getProperty(NatsConstants.NATS_STREAMING_CONNECT_WAIT);
+            String discoverPrefix = natsProperties.getProperty(NatsConstants.NATS_STREAMING_DISCOVER_PREFIX);
+            String maxPingsOut = natsProperties.getProperty(NatsConstants.NATS_STREAMING_MAX_PINGS_OUT);
+            String pingInterval = natsProperties.getProperty(NatsConstants.NATS_STREAMING_PING_INTERVAL);
+            String traceConnection = natsProperties.getProperty(NatsConstants.NATS_STREAMING_TRACE_CONNECTION);
+
+            Options.Builder builder = new Options.Builder().natsUrl(
+                    StringUtils.isEmpty(natsStreamingUrl) ? NatsConstants.DEFAULT_NATS_STREAMING_URL : natsStreamingUrl)
+                    .clientId(natsStreamingClientId).clusterId(StringUtils.isEmpty(natsStreamingClusterId) ?
+                            NatsConstants.DEFAULT_NATS_STREAMING_CLUSTER_ID :
+                            natsStreamingClusterId).connectionLostHandler((streamingConnection, e) -> {
+                        NatsMessageConsumer messageConsumer = NatsEndpointManager.getInstance().getMessageConsumer();
+                        try {
+                            if (connection != null) connection.close();
+                            connection = null;
+                            subscription = null;
+                            messageConsumer.consumeMessage();
```

Comment:
why are we calling this here? Isn't it introducing a cyclic dependency? We already call this method at NatsEndpointManager.startListener

Reply:
With NATS Streaming, once a connection is lost it is not recreated once the server is back up. So we need to recreate the connection again in the connectionlosthandler method.